test/e2e/test_helm_install_upgrade.go

```diff
@@ -0,0 +1,403 @@
+		image = "docker.io/kong/gateway-operator-oss"
+
+		waitTime = 3 * time.Minute
+	)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	// createEnvironment will queue up environment cleanup if necessary
+	// and dumping diagnostics if the test fails.
+	e := CreateEnvironment(t, ctx)
+
+	// assertion is run after the upgrade to assert the state of the resources in the cluster.
+	type assertion struct {
+		Name string
+		Func func(*assert.CollectT, *testutils.K8sClients)
+	}
+
+	testcases := []struct {
+		name             string
+		fromVersion      string
+		toVersion        string
+		objectsToDeploy  []client.Object
+		upgradeToCurrent bool
+		assertions       []assertion
+	}{
+		// NOTE: We do not support versions earlier than 1.2 with the helm chart.
+		// The initial version of the chart contained CRDs from KGO 1.2. which
+		// introduced a breaking change which makes it impossible to upgrade from
+		// automatically (without manually deleting the CRDs).
+		{
+			name:        "upgrade from 1.2.0 to 1.2.3",
+			fromVersion: "1.2.0",
+			toVersion:   "1.2.3",
+			objectsToDeploy: []client.Object{
+				&operatorv1beta1.GatewayConfiguration{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: "gwconf-upgrade-120-123",
+					},
+					Spec: baseGatewayConfigurationSpec(),
+				},
+				&gatewayv1.GatewayClass{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: "gwclass-upgrade-120-123",
+					},
+					Spec: gatewayv1.GatewayClassSpec{
+						ParametersRef: &gatewayv1.ParametersReference{
+							Group:     gatewayv1.Group(operatorv1beta1.SchemeGroupVersion.Group),
+							Kind:      gatewayv1.Kind("GatewayConfiguration"),
+							Namespace: (*gatewayv1.Namespace)(&e.Namespace.Name),
+							Name:      "gwconf-upgrade-120-123",
+						},
+						ControllerName: gatewayv1.GatewayController(vars.ControllerName()),
+					},
+				},
+				&gatewayv1.Gateway{
+					ObjectMeta: metav1.ObjectMeta{
+						GenerateName: "gw-upgrade-120-123-",
+						Labels: map[string]string{
+							"gw-upgrade-120-123": "true",
+						},
+					},
+					Spec: gatewayv1.GatewaySpec{
+						GatewayClassName: gatewayv1.ObjectName("gwclass-upgrade-120-123"),
+						Listeners: []gatewayv1.Listener{{
+							Name:     "http",
+							Protocol: gatewayv1.HTTPProtocolType,
+							Port:     gatewayv1.PortNumber(80),
+						}},
+					},
+				},
+			},
+			assertions: []assertion{
+				{
+					Name: "gateway is programmed",
+					Func: func(c *assert.CollectT, cl *testutils.K8sClients) {
+						gws, err := cl.GatewayClient.GatewayV1().Gateways(e.Namespace.Name).List(ctx, metav1.ListOptions{
+							LabelSelector: "gw-upgrade-120-123=true",
+						})
+						require.NoError(c, err)
+						require.Len(c, gws.Items, 1)
+						gw := &gws.Items[0]
+						assert.True(c, gateway.IsProgrammed(gw))
+						assert.True(c, gateway.AreListenersProgrammed(gw))
+					},
+				},
+			},
+		},
+		{
+			name:             "upgrade from 1.2.3 to current",
```

Comment:
We'll probably need a release manual/automated action that will enforce updating test cases to cover the latest patch -> current upgrade path. Can we add this to the release procedure as a manual step for now?

Reply:
Added a comment about renovate and https://github.com/Kong/gateway-operator/issues/121. PTAL

d153f0ae